File: Trees/trie.java
/*
Description
A trie (pronounced as "try") or prefix tree is a tree data structure used to efficiently store and retrieve keys in a dataset of strings. There are various applications of this data structure, such as autocomplete and spellchecker.

Implement the Trie class:

Trie() Initializes the trie object.
void insert(String word) Inserts the string word into the trie.
boolean search(String word) Returns true if the string word is in the trie (i.e., was inserted before), and false otherwise.
boolean startsWith(String prefix) Returns true if there is a previously inserted string word that has the prefix prefix, and false otherwise.

Input
["Trie", "insert", "search", "search", "startsWith", "insert", "search"]
[[], ["apple"], ["apple"], ["app"], ["app"], ["app"], ["app"]]
Output
[null, null, true, false, true, null, true]

Explanation
Trie trie = new Trie();
trie.insert("apple");
trie.search("apple");   // return True
trie.search("app");     // return False
trie.startsWith("app"); // return True
trie.insert("app");
trie.search("app");     // return True

Time Complexity for each operation is O(n)
Space Complexity O(n*m) where n is the number of words inserted and m is the average length of the words.

Explanation:
insert() -> traverse through each character of the input word and initializes it if necessary. If the end of the word is reached set isEnd to true.
search() -> Search In each child node until the end of the word is reached, then if end of the node is also reached return true else false.
startsWith() -> Similar to search method but we only check if end of the prefix is reached and we don't need to check if it is the end of the node.

*/

class Trie {
    Node root;

    public Trie() {
        root = new Node();
    }
    
    public void insert(String word) {
        root.insert(word, 0);
    }
    
    public boolean search(String word) {
        return root.search(word, 0);
    }
    
    public boolean startsWith(String prefix) {
        return root.startsWith(prefix, 0);
    }

    class Node {
        Node[] nodes;
        boolean isEnd;

        Node() {
            nodes = new Node[26];
        }

        // Function to insert the word in the tree
        private void insert(String word, int idx) {
            if (idx >= word.length()) return; // handle edge case
            int i = word.charAt(idx) - 'a';
            if (nodes[i] == null) {
                nodes[i] = new Node(); // initialize the node[i] if the letter was not found before
            }

            if (idx == word.length()-1) nodes[i].isEnd = true; // signifies that this is the end of the word
            nodes[i].insert(word, idx+1); // recursive call to populate the child node
        }

        // Function to search the word in the tree
        private boolean search(String word, int idx) {
            if (idx >= word.length()) return false;
            Node node = nodes[word.charAt(idx) - 'a'];
            if (node == null) return false; // if the node is null it means that it was not initialised hence the character was never found.
            if (idx == word.length() - 1 && node.isEnd) return true; //if it is the last character and the end of the node then return true

            return node.search(word, idx+1); // recursive call search in the child node

        }

        //Function to search the prefix in tree
        private boolean startsWith(String prefix, int idx) {
            if (idx >= prefix.length()) return false;
            Node node = nodes[prefix.charAt(idx) - 'a'];
            if (node == null) return false;
            if (idx == prefix.length() - 1) return true; // Very similar to above method but here we don't need to check if it is the end of the node.

            return node.startsWith(prefix, idx+1);
        }
    }
}

File: Trees/trie.go
/*
Approach:
In this implementation, we define two struct types: `TrieNode` and `Trie`. `TrieNode` represents a single node in the Trie. It has two fields: `children` (a map of rune to `TrieNode`) to store the child nodes and `isWord` (a boolean) to indicate the end of a word.

`Trie` represents the Trie data structure and has a single field `root` (a pointer to `TrieNode`), which is initialized with an empty `TrieNode` in the `NewTrie` function.

The Trie supports three operations:
- `Insert` function is used to add a word to the Trie. It starts from the root node and traverses through each character of the word, creating new nodes as needed.
- `Search` function is used to check if a given word exists in the Trie. It follows a similar approach to the `Insert` function but returns `true` if the final node marks the end of a word.
- `StartsWith` function is used to check if there is any word in the Trie that starts with a given prefix. It works similar to the `Search` function but does not require the final node to mark the end of a word.

Sample Input:
In the `main` function demonstrates the usage of the Trie data structure. We insert words "apple", "app", "application", "book", and "dog" into the Trie. Then, we search for words "app", "apple", and "banana". Finally, we check if there are any words in the Trie that start with prefixes "ap", "do", and "cat".

Time complexity of the `Insert`, `Search`, and `StartsWith` operations: O(m), where m is the length of the word or prefix being processed. 

Space complexity: O(n), where n is the total number of characters in all the words inserted in the Trie.
*/

//Here's an implementation of the Trie data structure in Go:

package main

import "fmt"

// TrieNode represents a single node in the Trie
type TrieNode struct {
	children map[rune]*TrieNode
	isWord   bool
}

// Trie represents the Trie data structure
type Trie struct {
	root *TrieNode
}

// NewTrie initializes a new Trie
func NewTrie() *Trie {
	return &Trie{
		root: &TrieNode{
			children: make(map[rune]*TrieNode),
			isWord:   false,
		},
	}
}

// Insert adds a word to the Trie
func (t *Trie) Insert(word string) {
	node := t.root

	// Traverse through each character of the word
	for _, ch := range word {
		if node.children[ch] == nil {
			node.children[ch] = &TrieNode{
				children: make(map[rune]*TrieNode),
				isWord:   false,
			}
		}
		node = node.children[ch]
	}

	// Mark the end of the word
	node.isWord = true
}

// Search returns true if the word exists in the Trie, otherwise false
func (t *Trie) Search(word string) bool {
	node := t.root

	// Traverse through each character of the word
	for _, ch := range word {
		if node.children[ch] == nil {
			return false
		}
		node = node.children[ch]
	}

	// Check if the node marks the end of a word
	return node.isWord
}

// StartsWith returns true if there is any word in the Trie that starts with the given prefix, otherwise false
func (t *Trie) StartsWith(prefix string) bool {
	node := t.root

	// Traverse through each character of the prefix
	for _, ch := range prefix {
		if node.children[ch] == nil {
			return false
		}
		node = node.children[ch]
	}

	return true
}

func main() {
	trie := NewTrie()

	// Inserting words into the Trie
	trie.Insert("apple")
	trie.Insert("app")
	trie.Insert("application")
	trie.Insert("book")
	trie.Insert("dog")

	// Searching words in the Trie
	fmt.Println(trie.Search("app"))    // Output: true
	fmt.Println(trie.Search("apple"))  // Output: true
	fmt.Println(trie.Search("banana")) // Output: false

	// Checking prefixes in the Trie
	fmt.Println(trie.StartsWith("ap"))  // Output: true
	fmt.Println(trie.StartsWith("do"))  // Output: true
	fmt.Println(trie.StartsWith("cat")) // Output: false
}

